File: server/src/models/Trade.ts
import mongoose, { Schema } from 'mongoose';
import { v4 as uuidv4 } from 'uuid';

export interface ITradeImage {
  id: string;
  filename: string;
  originalName: string;
  mimeType: string;
  size: number;
  caption: string;
  sortOrder: number;
  createdAt: string;
  driveFileId?: string;
}

export interface ITradeRuleEntry {
  ruleId: string;
  ruleName: string;
  followed: boolean;
}

export interface ITrade {
  _id: string;
  date: string;
  instrument: string;
  side: 'LONG' | 'SHORT';
  entryPrice: number;
  exitPrice: number;
  quantity: number;
  pnl: number;
  fees: number;
  notes: string;
  tags: string[];
  images: ITradeImage[];
  ruleChecklist: ITradeRuleEntry[];
  stopLoss: number;
  takeProfit: number;
  rating: number;
  disciplineScore: number;
  isMissedTrade: boolean;
  playbookId: string;
  reviewNotes: string;
  mistakes: string;
  lessons: string;
  userEmail: string;
  createdAt: string;
  updatedAt: string;
}

const TradeImageSchema = new Schema<ITradeImage>({
  id: { type: String, default: () => uuidv4() },
  filename: { type: String, required: true },
  originalName: { type: String, required: true },
  mimeType: { type: String, required: true },
  size: { type: Number, required: true },
  caption: { type: String, default: '' },
  sortOrder: { type: Number, default: 0 },
  createdAt: { type: String, default: () => new Date().toISOString() },
  driveFileId: { type: String, default: '' },
}, { _id: false });

const TradeRuleEntrySchema = new Schema<ITradeRuleEntry>({
  ruleId: { type: String, required: true },
  ruleName: { type: String, required: true },
  followed: { type: Boolean, required: true },
}, { _id: false });

const TradeSchema = new Schema<ITrade>({
  _id: { type: String, default: () => uuidv4() },
  date: { type: String, required: true, index: true },
  instrument: { type: String, required: true },
  side: { type: String, enum: ['LONG', 'SHORT'], required: true },
  entryPrice: { type: Number, default: 0 },
  exitPrice: { type: Number, default: 0 },
  quantity: { type: Number, default: 0 },
  pnl: { type: Number, default: 0 },
  fees: { type: Number, default: 0 },
  notes: { type: String, default: '' },
  tags: { type: [String], default: [] },
  images: { type: [TradeImageSchema], default: [] },
  ruleChecklist: { type: [TradeRuleEntrySchema], default: [] },
  stopLoss: { type: Number, default: 0 },
  takeProfit: { type: Number, default: 0 },
  rating: { type: Number, default: 0 },
  disciplineScore: { type: Number, default: 0 },
  isMissedTrade: { type: Boolean, default: false },
  playbookId: { type: String, default: '' },
  reviewNotes: { type: String, default: '' },
  mistakes: { type: String, default: '' },
  lessons: { type: String, default: '' },
  userEmail: { type: String, required: true, index: true },
  createdAt: { type: String, default: () => new Date().toISOString() },
  updatedAt: { type: String, default: () => new Date().toISOString() },
});

export const Trade = mongoose.model<ITrade>('Trade', TradeSchema);

File: server-java/src/main/java/com/conarum/tradingjournal/domain/trade/model/Trade.java
package com.conarum.tradingjournal.domain.trade.model;

import com.conarum.tradingjournal.common.model.BaseEntity;
import lombok.Getter;
import lombok.Setter;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Getter
@Setter
@Document(collection = "trades")
@CompoundIndex(name = "user_date_idx", def = "{'userEmail': 1, 'date': -1}")
public class Trade extends BaseEntity {
    private String date; // Format: YYYY-MM-DD
    private String instrument;
    private String side; // LONG, SHORT
    private BigDecimal entryPrice = BigDecimal.ZERO;
    private BigDecimal exitPrice = BigDecimal.ZERO;
    private BigDecimal quantity = BigDecimal.ZERO;
    private BigDecimal pnl = BigDecimal.ZERO;
    private BigDecimal fees = BigDecimal.ZERO;
    private String notes = "";
    private List<String> tags = new ArrayList<>();
    private List<TradeImage> images = new ArrayList<>();
    private List<TradeRuleEntry> ruleChecklist = new ArrayList<>();
    private BigDecimal stopLoss = BigDecimal.ZERO;
    private BigDecimal takeProfit = BigDecimal.ZERO;
    private int rating = 0;
    private int disciplineScore = 0;
    private boolean isMissedTrade = false;
    @Indexed
    private String playbookId = "";
    private String reviewNotes = "";
    private String mistakes = "";
    private String lessons = "";

    @Getter
    @Setter
    public static class TradeImage {
        private String id = UUID.randomUUID().toString();
        private String filename;
        private String originalName;
        private String mimeType;
        private long size;
        private String caption = "";
        private int sortOrder = 0;
        private String createdAt;
        private String driveFileId = "";
    }

    @Getter
    @Setter
    public static class TradeRuleEntry {
        private String ruleId;
        private String ruleName;
        private boolean followed;
    }
}
